StandaloneMmPkg/Drivers/StandaloneMmIplPei/MmFoundationHob.c

```diff
@@ -379,6 +385,274 @@ MmIplBuildSmmProfileHobs (
+  )
+{
+  UINTN                 Index;
+  UINTN                 Count;
+  UINTN                 PlatformRegionCount;
+  UINTN                 UsedSize;
+  UINT64                PreviousAddress;
+  UINT64                MaxAddress;
+  MM_IPL_MEMORY_REGION  *MemoryRegions;
+  MM_IPL_MEMORY_REGION  SortBuffer;
+
+  MaxAddress = LShiftU64 (1, MmIplCalculateMaximumSupportAddress ());
+
+  //
+  // Get the count of platform memory regions
+  //
+  PlatformRegionCount = 0;
+  if ((PlatformHobList != NULL) && (PlatformHobSize != 0)) {
+    CollectPlatformMemoryRegions (PlatformHobList, PlatformHobSize, NULL, &PlatformRegionCount);
+  }
+
+  //
+  // Allocate buffer for platform memory regions, SmmProfile data, MMRam ranges, an extra terminator.
+  //
+  Count         = PlatformRegionCount + Block->NumberOfMmReservedRegions + ((SmmProfileDataHob != NULL) ? 1 : 0) + 1;
+  MemoryRegions = AllocatePages (EFI_SIZE_TO_PAGES (Count * sizeof (*MemoryRegions)));
+  ASSERT (MemoryRegions != NULL);
+  if (MemoryRegions == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a:%d - No enough memory\n", __func__, __LINE__));
+    CpuDeadLoop ();
+    return;
+  }
+
+  //
+  // The very last region is the terminator
+  //
+  MemoryRegions[Count - 1].Base   = MaxAddress;
+  MemoryRegions[Count - 1].Length = 0;
+
+  //
+  // Collect platform memory regions
+  //
+  if (PlatformRegionCount != 0) {
+    CollectPlatformMemoryRegions (PlatformHobList, PlatformHobSize, MemoryRegions, &PlatformRegionCount);
+  }
+
+  //
+  // Collect SMRAM regions
+  //
+  for (Index = 0; Index < Block->NumberOfMmReservedRegions; Index++) {
+    MemoryRegions[PlatformRegionCount + Index].Base   = Block->Descriptor[Index].CpuStart;
+    MemoryRegions[PlatformRegionCount + Index].Length = Block->Descriptor[Index].PhysicalSize;
+  }
+
+  //
+  // Collect SMM profile database region
+  //
+  if (SmmProfileDataHob != NULL) {
+    MemoryRegions[PlatformRegionCount + Block->NumberOfMmReservedRegions].Base   = SmmProfileDataHob->AllocDescriptor.MemoryBaseAddress;
+    MemoryRegions[PlatformRegionCount + Block->NumberOfMmReservedRegions].Length = SmmProfileDataHob->AllocDescriptor.MemoryLength;
+  }
+
+  //
+  // Build system memory resource HOBs excluding platform memory regions, SMRAM regions, SmmProfile database.
+  //
+  QuickSort (MemoryRegions, Count, sizeof (*MemoryRegions), MemoryRegionBaseAddressCompare, &SortBuffer);
+  UsedSize        = 0;
+  PreviousAddress = 0;
+  for (Index = 0; Index < Count; Index++) {
+    ASSERT (MaxAddress >= MemoryRegions[Index].Base + MemoryRegions[Index].Length);
+
+    if (MemoryRegions[Index].Base > PreviousAddress) {
+      if (*HobBufferSize >= UsedSize + sizeof (EFI_HOB_RESOURCE_DESCRIPTOR)) {
+        MmIplBuildMemoryResourceHob (
+          (EFI_HOB_RESOURCE_DESCRIPTOR *)(HobBuffer + UsedSize),
+          EFI_RESOURCE_SYSTEM_MEMORY,
+          PcdGetBool (PcdCpuSmmProfileEnable) ? EDKII_MM_RESOURCE_ATTRIBUTE_LOGGING : 0,

```

Comment:
Add comments to explain the access to all system memory is logged by MmCpu driver.